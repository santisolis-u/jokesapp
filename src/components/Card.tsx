import React from 'react';
import {Animated, StyleSheet, Text, Dimensions} from 'react-native';

const {width} = Dimensions.get('window');

interface JokeCardProps {
  text: string;
  backgroundColor: string;
}

const JokeCard = ({text, backgroundColor}: JokeCardProps) => {
  return (
    <Animated.View style={[styles.card, {backgroundColor}]}>
      <Text style={[styles.text]}>{text}</Text>
      <Text style={[styles.author]}>@designhumor</Text>
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  card: {
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingTop: 70,
    paddingBottom: 30,
    paddingHorizontal: 20,
    borderRadius: 8,
    marginHorizontal: 20,
    width: width * 0.8,
  },
  text: {
    color: 'white',
    textAlign: 'center',
    fontSize: 24,
    marginHorizontal: 20,
    fontFamily: 'OpenSans-Bold',
  },
  author: {
    fontSize: 22,
    color: 'white',
    fontFamily: 'IndieFlower',
  },
  bottom: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  authorWithIcon: {
    marginRight: 8,
  },
});

export default JokeCard;
