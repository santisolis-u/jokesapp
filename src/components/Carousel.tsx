import {Animated, Dimensions, StyleSheet, Text, View} from 'react-native';
import React from 'react';
import Card from './Card';
import Indicator from './Indicator';
import {Joke} from '../hooks/useJokes';
import {JOKE_CARD_COLORS} from '../../utils/Colors';

const {width} = Dimensions.get('window');

interface CarouselProps {
  items: Joke[];
  indexChange: (number: number) => void;
  fetchNextJokes: () => void;
}

const Carousel = ({items, indexChange, fetchNextJokes}: CarouselProps) => {
  const scrollX = React.useRef(new Animated.Value(0)).current;
  return (
    <View style={styles.container}>
      <Animated.FlatList
        data={items}
        scrollEventThrottle={32}
        pagingEnabled
        onScroll={Animated.event(
          [
            {
              nativeEvent: {contentOffset: {x: scrollX}},
            },
          ],
          {
            useNativeDriver: false,
            listener: ({nativeEvent}) => {
              const index = Math.ceil(
                (nativeEvent as any).contentOffset.x / width,
              );
              if (indexChange) {
                indexChange(index);
              }
            },
          },
        )}
        keyExtractor={element => element.id}
        renderItem={item => (
          <Card
            backgroundColor={
              JOKE_CARD_COLORS[
                Math.floor(
                  Math.random() * (JOKE_CARD_COLORS.length - 1 + 1) + 0,
                )
              ]
            }
            text={item.item.joke}
          />
        )}
        horizontal
        showsHorizontalScrollIndicator={false}
        onEndReached={() => {
          if (fetchNextJokes) {
            fetchNextJokes();
          }
        }}
      />
      <Indicator itemsQty={items.length} scrollX={scrollX} />
    </View>
  );
};

export default Carousel;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginHorizontal: width * 0.05,
  },
});
