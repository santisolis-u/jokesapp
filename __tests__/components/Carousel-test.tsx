import 'react-native';
import React from 'react';

import Card from '../../src/components/Card';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import Carousel from '../../src/components/Carousel';
import {fireEvent, render} from '@testing-library/react-native';

const indexChange = jest.fn();
const fetchNextJokes = jest.fn();
const items = [
  {
    id: '1',
    joke: 'Test',
    permalink: '1',
  },
];
describe('Renders', () => {
  it('Renders correctly', () => {
    const json = renderer
      .create(
        <Carousel
          items={items}
          indexChange={indexChange}
          fetchNextJokes={fetchNextJokes}
          cardIndex={0}
          renderItem={jest.fn()}
        />,
      )
      .toJSON();
    expect(json).toMatchSnapshot();
  });
  it('onFetchMore gets invoked when end reached', async () => {
    const {getByTestId} = render(
      <Carousel
        items={items}
        indexChange={indexChange}
        fetchNextJokes={fetchNextJokes}
        cardIndex={0}
        renderItem={jest.fn()}
      />,
    );
    const toScroll = await getByTestId('flatList');
    fireEvent.scroll(toScroll, {
      nativeEvent: {
        contentOffset: {
          y: 0,
          x: 500,
        },
        contentSize: {
          // Dimensions of the scrollable content
          height: 100,
          width: 500,
        },
        layoutMeasurement: {
          // Dimensions of the device
          height: 100,
          width: 100,
        },
      },
    });
    expect(fetchNextJokes).toHaveBeenCalledTimes(1);
  });
});
