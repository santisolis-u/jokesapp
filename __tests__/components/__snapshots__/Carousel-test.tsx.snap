// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "marginHorizontal": 37.5,
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    collapsable={false}
    data={
      Array [
        Object {
          "id": "1",
          "joke": "Test",
          "permalink": "1",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={[Function]}
    pagingEnabled={true}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={64}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={Object {}}
    testID="flatList"
    viewabilityConfig={
      Object {
        "itemVisiblePercentThreshold": 90,
        "minimumViewTime": 5,
        "waitForInteraction": true,
      }
    }
    viewabilityConfigCallbackPairs={
      Array [
        Object {
          "onViewableItemsChanged": [Function],
          "viewabilityConfig": Object {
            "itemVisiblePercentThreshold": 90,
            "minimumViewTime": 5,
            "waitForInteraction": true,
          },
        },
      ]
    }
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      />
    </View>
  </RCTScrollView>
  <View
    collapsable={false}
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 32,
      }
    }
  >
    <View
      style={
        Object {
          "maxWidth": 712.5,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "black",
            "borderRadius": 8,
            "height": 5,
            "marginHorizontal": 8,
            "opacity": 1,
            "width": 50,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "maxWidth": 712.5,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "black",
            "borderRadius": 8,
            "height": 5,
            "marginHorizontal": 8,
            "opacity": 0.1,
            "width": 50,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "maxWidth": 712.5,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "black",
            "borderRadius": 8,
            "height": 5,
            "marginHorizontal": 8,
            "opacity": 0.1,
            "width": 50,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "maxWidth": 712.5,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "black",
            "borderRadius": 8,
            "height": 5,
            "marginHorizontal": 8,
            "opacity": 0.1,
            "width": 50,
          }
        }
      />
    </View>
  </View>
</View>
`;
